utils/cautils/token_generator.go

```diff
@@ -77,24 +86,187 @@ func (t *TokenGenerator) Token(sub string, opts ...token.Options) (string, error
+	if ctx.Bool("console") {
+		args = append(args, "--console")
+	}
+	if p.ListenAddress != "" {
+		args = append(args, "--listen", p.ListenAddress)
+	}
+	out, err := exec.Step(args...)
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(out)), nil
+}
+
+type tokenAttrs struct {
+	subject                     string
+	root                        string
+	caURL                       string
+	audience                    string
+	issuer                      string
+	kid                         string
+	sans                        []string
+	notBefore, notAfter         time.Time
+	certNotBefore, certNotAfter provisioner.TimeDuration
+}
+
+func generateX5CToken(ctx *cli.Context, p *provisioner.X5C, tokType int, tokAttrs tokenAttrs) (string, error) {
+	x5cCertFile := ctx.String("x5c-cert")
+	x5cKeyFile := ctx.String("x5c-key")
+	if len(x5cCertFile) == 0 {
+		return "", errs.RequiredWithProvisionerTypeFlag(ctx, "X5C", "x5c-cert")
+	}
+	if len(x5cKeyFile) == 0 {
+		return "", errs.RequiredWithProvisionerTypeFlag(ctx, "X5C", "x5c-key")
+	}
+
+	// Get private key from given key file
+	var opts []jose.Option
+	if passwordFile := ctx.String("password-file"); len(passwordFile) != 0 {
+		opts = append(opts, jose.WithPasswordFile(passwordFile))
+	}
+	jwk, err := jose.ParseKey(x5cKeyFile, opts...)
+	if err != nil {
+		return "", err
+	}
+	tokenGen := NewTokenGenerator(jwk.KeyID, p.Name,
+		fmt.Sprintf("%s#%s", tokAttrs.audience, p.GetID()), tokAttrs.root,
+		tokAttrs.notBefore, tokAttrs.notAfter, jwk)
+	switch tokType {
+	case SignType:
+		return tokenGen.SignToken(tokAttrs.subject, tokAttrs.sans, token.WithX5CFile(x5cCertFile, jwk.Key))
+	case RevokeType:
+		return tokenGen.RevokeToken(tokAttrs.subject, token.WithX5CFile(x5cCertFile, jwk.Key))
+	case SSHUserSignType:
+		return tokenGen.SignSSHToken(tokAttrs.subject, provisioner.SSHUserCert, tokAttrs.sans,
+			tokAttrs.certNotBefore, tokAttrs.certNotAfter, token.WithX5CFile(x5cCertFile, jwk.Key))
+	case SSHHostSignType:
+		return tokenGen.SignSSHToken(tokAttrs.subject, provisioner.SSHHostCert, tokAttrs.sans,
+			tokAttrs.certNotBefore, tokAttrs.certNotAfter, token.WithX5CFile(x5cCertFile, jwk.Key))
+	default:
+		return tokenGen.Token(tokAttrs.subject, token.WithX5CFile(x5cCertFile, jwk.Key))
+	}
+}
+
+// loadJWK loads a JWK from a combination of the
+func loadJWK(ctx *cli.Context, p *provisioner.JWK, tokAttrs tokenAttrs) (jwk *jose.JSONWebKey, kid string, err error) {
+	var opts []jose.Option
+	if passwordFile := ctx.String("password-file"); len(passwordFile) != 0 {
+		opts = append(opts, jose.WithPasswordFile(passwordFile))
+	}
+
+	if keyFile := ctx.String("key"); len(keyFile) == 0 {
+		if p == nil {
+			return nil, "", errors.New("offline mode with no CA configuration file must specify a JWK file (--key)")
+		}
+		kid = p.Key.KeyID
+		// If provisioner is not nil then we must be using the offlineCA.
+		var encryptedKey string
+		if ctx.IsSet("offline") {
```

Comment:
It should be `ctx.Bool("offline")`, even if it's not common `--offline=false` is possible